### PR TITLE
Implementation of rectangular members in HD and SD

### PR DESCRIPTION
**Pull request by luwang00**
**Feature or improvement description**
This PR adds the implementation of rectangular members in HydroDyn and SubDyn. It also includes various other bug fixes and improvements including the following:

_HydroDyn_
* In HydroDyn, ballast water is now handled consistently assuming no internal free surface when computing the inertial loads and the internal hydrostatic loads. This is consistent with the fact that HydroDyn does not model internal sloshing.
* In HydroDyn, corrections were made to the rotational equations of motion of ballast water and marine growth.
* In HydroDyn, inertial loads from marine growth on joints were previously zeroed when the joint emerges from water. This is now corrected.

_SubDyn_
* In SubDyn, the implementation of arbitrary-section beams was previously incomplete. This is addressed by this PR.
* In SubDyn, it was previously assumed that the torsion constant is the same as the section polar area moment of inertia for arbitrary beam sections. However, this is only true for circular sections. Now, two separate values can be provided for the torsion constant and polar area moment of inertia for arbitrary beam sections.
* In SubDyn, the orientation of arbitrary-section beams was previously set according to user-defined direction cosine matrices. However, there was no check on whether the user-defined direction cosine matrices are valid rotation matrices, or that they are consistent with the member end joints. Now the orientation of arbitrary-section beams is determined based on its two end joints plus a spin angle about its axis, same as rectangular members. The spin angle is also introduced for circular members to give users more control on the element local coordinate system for load outputs. Now, only spring elements still use the user-defined direction cosine matrices.
* In SubDyn, addressed Issue #2634 by checking the orthogonality and the determinants of user-defined direction cosine matrices.
* SubDyn now errors out if the `COSMID` of a spring member references a nonexistent entry in the direction cosine matrix table. Previously, it defaults back to the identity matrix by a series of coincidence without reporting an error.

_OpenFAST_IO_
* OpenFAST_IO has been updated with the new HydroDyn and SubDyn inputs.

**TO-DOs**
The HydroDyn strip-theory member visualization should be updated to support rectangular members. @andrew-platt and @deslaughter will help with this.

**Related issue, if one exists**
Issue #2634 

**Impacted areas of the software**
HydroDyn, SubDyn, OpenFAST_IO

**Test results, if applicable**
* Due to the changes and bug fixes to ballast-water and marine-growth modeling, results of r-test cases involving either need to be updated.
* The HydroDyn and SubDyn input files also need to be updated due to the addition of rectangular members. These are completed in the referenced r-test commit, and all tests passed.
* Added a new r-test for the semi-numerical computation of hydrostatic loads on rectangular members. 
* A new r-test with a cantilever rectangular-section beam in SubDyn is developed by @RBergua and added through this PR.

### CODE REVIEW

**Comment by RBergua** [comment]
I performed one verification for a cantilever beam using rectangular cross sections. This verification test is based on the OC7 Phase II project (undergoing). 

The first part of the verification is a static deflection test according to the gravity acceleration (9.80665 m/s^2) and a vertical force Fz = 1.345E7 N. The system and boundary conditions are as follows:
image

The properties of the flexible pontoon are:
image

The properties of the outer column (lumped mass) are:
image

I compared the results from SubDyn (linear assumption) against SAMCEF (non-linear approach) and HOTINT (non-linear approach). As can be observed, there is excellent agreement between the three codes for the vertical deflections along the pontoon:  
image

In addition, another verification regarding the eigenfrequencies of the system was performed. The below table compares the first 4 eigenfrequencies. Once again, very good agreement between SubDyn and SAMCEF is obtained:
image

This verification test has been added in the r-test repository: 
https://github.com/OpenFAST/r-test/tree/main/modules/subdyn/SD_CantileverBeam_Rectangular

For reference, a comparison between the Timoshenko beam theory (see above results) and the Euler-Bernoulli beam theory is provided next. As can be observed, when modeling the system with Euler-Bernoulli beams, the system behaves slightly stiffer (as expected). The vertical deflection of the last point when using Euler-Bernoulli beams is 2.5% smaller. Good agreement is observed between SubDyn and Project Chrono for the Euler-Bernoulli approach.
image

**Comment by andrew-platt** [comment]
- [ ] add visualization of rectangular members

**Comment by andrew-platt** [comment]
Visualization will be in a separate PR (requires changes to the meshes)

**Review by andrew-platt — APPROVED** [review]
Looks good.  I didn't go through all the equations, but the modifications look reasonable.

**Review by deslaughter — APPROVED** [review]
I think it looks good. We should get together sometime and work in restructuring Morison for better performance, but this is relatively easy to read. Just had a couple of suggestions to make the code easier to maintain.

File: modules/hydrodyn/src/HydroDyn_Input.f90 [hunk 1]
```diff
@@ -1854,99 +2157,312 @@ SUBROUTINE HydroDynInput_ProcessInitData( InitInp, Interval, InputFileData, ErrS
...
-            IF ( InputFileData%Morison%CoefMembers(I)%MemberID == InputFileData%Morison%CoefMembers(J)%MemberID ) THEN
-               CALL SetErrStat( ErrID_Fatal,'Duplicate MemberIDs were found in the Member-based Hydrodynamic coefficients table.',ErrStat,ErrMsg,RoutineName)
+         DO J = I+1,InputFileData%Morison%NCoefMembersCyl
+            IF ( InputFileData%Morison%CoefMembersCyl(I)%MemberID == InputFileData%Morison%CoefMembersCyl(J)%MemberID ) THEN
+               CALL SetErrStat( ErrID_Fatal,'Duplicate MemberIDs were found in the Member-based Hydrodynamic coefficients table for cylindrical members.',ErrStat,ErrMsg,RoutineName)
                RETURN
             END IF
          END DO
 
+         IF ( InputFileData%Morison%CoefMembersCyl(I)%MemberCd1 < 0 ) THEN
```
[the remark this answers is not in the document]
Comment (luwang00) [reply modules/hydrodyn/src/HydroDyn_Input.f90:2186]: Thanks for the suggestion. These checks and other similar instances are now cleaned up as suggested.

File: modules/hydrodyn/src/Morison.f90 [hunk 2]
```diff
@@ -550,7 +730,7 @@ SUBROUTINE WriteSummaryFile( UnSum, numJoints, numNodes, nodes, numMembers, memb
    real(ReKi)                                  :: ptLoad(6)
    logical                                     :: fillFlag
    type(Morison_MemberType)                    :: mem
-   REAL(ReKi)                                  :: Cd1, Cd2, Ca1, Ca2, Cp1, Cp2, AxCd1, AxCd2, AxCa1, AxCa2, AxCp1, AxCp2, Cb1, Cb2, JAxCd1, JAxCd2, JAxCa1, JAxCa2, JAxCp1, JAxCp2 ! tmp coefs
+   REAL(ReKi)                                  :: Cd1, CdA1, CdB1, Cd2, CdA2, CdB2, Ca1, CaA1, CaB1, Ca2, CaA2, CaB2, Cp1, Cp2, AxCd1, AxCd2, AxCa1, AxCa2, AxCp1, AxCp2, Cb1, Cb2, JAxCd1, JAxCd2, JAxCa1, JAxCa2, JAxCp1, JAxCp2 ! tmp coefs
```
[the remark this answers is not in the document]
Comment (luwang00) [reply modules/hydrodyn/src/Morison.f90]: Done!

File: modules/hydrodyn/src/Morison.f90 [hunk 3]
```diff
@@ -781,15 +961,15 @@ SUBROUTINE WriteSummaryFile( UnSum, numJoints, numNodes, nodes, numMembers, memb
...
+      WRITE( UnSum, '(1X,A5,31(2X,A10))' ) '  i  ', '  MbrIndx ', '   Nxi    ', '   Nyi    ', '   Nzi    ', '  Shape   ', '     R    ', '    SA    ', '    SB    ', '    t     ', '   tMG    ', '  MGDens  ', ' PropPot  ', 'FilledFlag', 'FilledMass', '    Cd    ', '    CdA   ', '    CdB   ', '    Ca    ', '    CaA   ', '    CaB   ', '    Cp    ', '    Cb    ', '   AxCd   ',  '   AxCa   ', '   AxCp   ', '   JAxCd  ', '   JAxCa  ', '   JAxCp  ', ' JAxFDMod ', 'JAxVnCOff ', 'JAxFDLoFSc'
+      WRITE( UnSum, '(1X,A5,31(2X,A10))' ) ' (-) ', '    (-)   ', '   (m)    ', '   (m)    ', '   (m)    ', '   (-)    ', '    (m)   ', '    (m)   ', '    (m)   ', '   (m)    ', '   (m)    ', ' (kg/m^3) ', '   (-)    ', '   (-)    ', '  (kg)    ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ',  '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   ', '    (-)   '
```
[the remark this answers is not in the document]
Comment (luwang00) [reply modules/hydrodyn/src/Morison.f90]: Done!

File: modules/hydrodyn/src/Morison.f90 [hunk 4]
```diff
@@ -781,15 +961,15 @@ SUBROUTINE WriteSummaryFile( UnSum, numJoints, numNodes, nodes, numMembers, memb
...
          ! need to add MSL2SWL offset from this because the Positons are relative to SWL, but we should report them relative to MSL here
          pos = nodes(i)%Position
          pos(3) = pos(3) + p%WaveField%MSL2SWL
-         write( UnSum, '(1X,I5,(2X,A10),3(2X,F10.4),2(2X,A10),2(2X,ES10.3),10(2X,A10),3(2X,ES10.3))' ) i,'    -     ', pos, '    -     ',  '    -     ',  nodes(i)%tMG,  nodes(i)%MGdensity,  '    -     ',  '    -     ',  '    -     ', '    -     ',  '    -     ',  '    -     ',  '    -     ',  '    -     ',  '    -     ',  '    -     ',  nodes(i)%JAxCd,  nodes(i)%JAxCa, nodes(i)%JAxCp
+         write( UnSum, '(1X,I5,(2X,A10),3(2X,F10.4),5(2X,A10),2(2X,ES10.3),14(2X,A10),3(2X,ES10.3),1X,I10,2(2X,ES10.3))' ) i,'    -     ', pos, '    -     ', '    -     ', '    -     ', '    -     ',  '    -     ',  nodes(i)%tMG,  nodes(i)%MGdensity,  '    -     ',  '    -     ',  '    -     ', '    -     ', '    -     ', '    -     ',  '    -     ', '    -     ', '    -     ', '    -     ',  '    -     ',  '    -     ',  '    -     ',  '    -     ',  nodes(i)%JAxCd,  nodes(i)%JAxCa, nodes(i)%JAxCp, nodes(i)%JAxFDMod, nodes(i)%JAxVnCOff, nodes(i)%JAxFDLoFSc
```
[the remark this answers is not in the document]
Comment (luwang00) [reply modules/hydrodyn/src/Morison.f90]: Done!

File: modules/hydrodyn/src/Morison.f90 [hunk 5]
```diff
@@ -808,16 +988,20 @@ SUBROUTINE WriteSummaryFile( UnSum, numJoints, numNodes, nodes, numMembers, memb
...
+      write( UnSum, '(1X,A8,2X,A6,2X,A6,45(2X,A12))' ) '  (-)   ', ' (-)  ',' (-)  ','   (m)    ', '    (-)     ', '   (m^3)    ', '   (m^3)    ', '      (m)   ', '     (m)    ', '     (m)    ', '     (m)    ', '      (m)   ', '     (m)    ', '     (m)    ', '     (m)    ', '   (-)    ', '   (-)    ', ' (kg/m^3)  ', '     (-)    ', '    (kg)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)   ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '    (-)     ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    ', '     (-)    '
```
[the remark this answers is not in the document]
Comment (luwang00) [reply modules/hydrodyn/src/Morison.f90]: Done!

File: modules/hydrodyn/src/Morison.f90 [hunk 6]
```diff
@@ -1383,6 +1791,52 @@ subroutine AllocateMemberDataArrays( member, memberLoads, errStat, errMsg )
    memberLoads%F_WMG    = 0.0_ReKi
    memberLoads%F_IMG    = 0.0_ReKi
 
+   if (member%MSecGeom == MSecGeom_Cyl) then
```
[the remark this answers is not in the document]
Comment (luwang00) [reply modules/hydrodyn/src/Morison.f90:1804]: Let's keep these for now. We can remove the zeroing in the future.
